olive/passes/onnx/moe_experts_distributor.py

```diff
@@ -0,0 +1,259 @@
+            output_name = node_name + "_output_0"
+            alltoall = onnx.helper.make_node(
+                "AllToAll", prev_node.output, [output_name], name=node_name, domain=_domain, group_size=world_size
+            )
+
+            model.add_node(alltoall)
+            OnnxModel.replace_node_input(next_node, prev_node.output[0], output_name)
+
+
+def _replace_constant_value(constant: Message, value: int):
+    for attr in constant.attribute:
+        if attr.name == "value":
+            attr.CopyFrom(
+                onnx.helper.make_attribute(
+                    attr.name, onnx.numpy_helper.from_array(numpy.array([value], numpy.int64)), attr.doc_string
+                )
+            )
+            return
+
+
+def _fix_shapes(
+    model: OnnxModel,
+    nodes: Dict[str, Message],
+    producers: Dict[str, Message],
+    consumers: Dict[str, Message],
+    experts: List[List[str]],
+    world_size: int,
+    num_experts: int,
+    rank: int,
+):
+    experts_per_rank = int(num_experts / world_size)
+
+    for expert in experts:
+        div_node = nodes[expert[5]]
+        div_node_iname = div_node.input[1]
+        div_node_oname = div_node.output[0]
+        _replace_constant_value(producers[div_node_iname], experts_per_rank)
+
+        add_node = nodes[expert[4]]
+        add_node_iname = add_node.input[1]
+        _replace_constant_value(producers[add_node_iname], experts_per_rank - 1)
+
+        mul_nodes = [n for n in consumers[div_node_oname] if n.op_type == "Mul"]
+        index = 1
+        for mul_node in mul_nodes:
+            if div_node_oname in mul_node.input:
+                _replace_constant_value(producers[mul_node.input[1]], index)
+                index += 1
+
+        reshape_node = nodes[expert[1]]
+        for parent in model.get_parents(reshape_node, producers):
+            if parent.op_type == "Concat":
+                concat_iname_0 = parent.input[0]
+                _replace_constant_value(producers[concat_iname_0], int(num_experts / experts_per_rank))
+
+                concat_iname_1 = parent.input[1]
+                _replace_constant_value(producers[concat_iname_1], experts_per_rank)
+
+                break
+
+
+def run(world_size: int, input_filepath: str, output_dirpath: str, debug: bool = False):
+    basename = Path(input_filepath).stem
+    output_dirpath = Path(output_dirpath)
+
+    model = OnnxModel(onnx.load_model(input_filepath))
+    producers = model.output_name_to_node()
+    consumers = model.input_name_to_nodes()
+
+    if debug:
+        OnnxModel.graph_topological_sort(model.model.graph)
+        _dump_graph(model.model, str(output_dirpath / f"{basename}.graph"))
+
+    experts = []
+    num_experts = None
+    for expert in _iterate_experts(model, producers):
+        if debug:
+            pprint.pprint([n.name for n in expert])
+
+        div_node = expert[5]
+        div_node_oname = div_node.output[0]
```

Comment:
What is so special about 5th expert ?

Reply:
It's not the 5th expert. It's the 5th node of every expert which happens to be the DIV node. Indexing is based on what was matched to identify the experts.